File: output_20260318_123118/claude-opus/c/Stack/filter.c
#include "stdlib.h"

struct node
{
    struct node *next;
    int value;
};

/*@ predicate nodes(struct node *n; list<int> vs) =
      n == 0 ?
        vs == nil
      :
        n->value |-> ?v &*& n->next |-> ?next &*& malloc_block_node(n) &*&
        nodes(next, ?rest) &*& vs == cons(v, rest);
@*/

struct stack
{
    struct node *head;
};

/*@ predicate stack(struct stack *stack; list<int> contents) =
      stack->head |-> ?head &*& malloc_block_stack(stack) &*& nodes(head, contents);
@*/

struct stack *create_stack()
/*@ requires true; @*/
/*@ ensures stack(result, nil); @*/
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    
    return stack;
}

void stack_push(struct stack *stack, int value)
/*@ requires stack(stack, ?contents); @*/
/*@ ensures stack(stack, cons(value, contents)); @*/
{
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    n->next = stack->head;
    n->value = value;
    stack->head = n;
}

int stack_pop(struct stack *stack)
/*@ requires stack(stack, ?contents) &*& contents != nil; @*/
/*@ ensures stack(stack, tail(contents)) &*& result == head(contents); @*/
{
    struct node *head = stack->head;
    
    int result = head->value;
    stack->head = head->next;
    free(head);
    
    return result;
}

typedef bool int_predicate(int x);

/*@ 
fixpoint bool pred_applies(int_predicate *p, int x);

fixpoint list<int> filter_list(int_predicate *p, list<int> xs) {
  switch(xs) {
    case nil: return nil;
    case cons(x, xs0): return pred_applies(p, x) ? cons(x, filter_list(p, xs0)) : filter_list(p, xs0);
  }
} 
@*/

/*@ 
predicate nodes_filter_pred(struct node *n; list<int> vs, int_predicate *p) =
    n == 0 ? vs == nil
    : exists<int> v, struct node *next;
      n->value |-> v &*& n->next |-> next &*& malloc_block_node(n) &*&
      nodes_filter_pred(next, ?rest, p) &*&
      vs == (pred_applies(p, v) ? cons(v, rest) : rest);
@*/

/*@ 
lemma void nodes_filter_relate(struct node *n, list<int> vs, int_predicate *p)
  requires nodes(n, vs);
  ensures nodes_filter_pred(n, filter_list(p, vs), p);
{
  switch(n) {
    case 0:
    case null: return;
    default:
      open nodes(n, vs);
      open nodes(n->next, ?rest);
      nodes_filter_relate(n->next, rest, p);
      close nodes_filter_pred(n, filter_list(p, vs), p);
      close nodes(n, vs);
  }
}
@*/

struct node *nodes_filter(struct node *n, int_predicate *p)
/*@ requires nodes(n, ?vs) &*& int_predicate(p); @*/
/*@ ensures nodes_filter_pred(result, filter_list(p, vs), p); @*/
{
    if (n == 0)
    {
        return 0;
    }
    else
    {
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            
            n->next = next;
            
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
/*@ requires stack(stack, ?contents) &*& int_predicate(p); @*/
/*@ ensures stack(stack, filter_list(p, contents)); @*/
{
    struct node *head = nodes_filter(stack->head, p);
    stack->head = head;
}

void nodes_dispose(struct node *n)
/*@ requires nodes(n, ?vs); @*/
/*@ ensures true; @*/
{
    if (n != 0)
    {
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
/*@ requires stack(stack, ?contents); @*/
/*@ ensures true; @*/
{
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x)
/*@ requires true; @*/
/*@ ensures true; @*/
{
    return x != 20;
}

int main()
/*@ requires true; @*/
/*@ ensures true; @*/
{
    struct stack *s = create_stack();
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    stack_filter(s, neq_20);
    stack_dispose(s);
    return 0;
}

File: output_20260318_123118/claude-opus/rust/Stack/filter.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

predicate node(struct Node *n; int v, struct Node *next) =
    n->value |-> v &*& n->next |-> next;

predicate nodes(struct Node *n; list<int> vs) =
    n == 0 ? vs == nil :
    node(n, head(vs), ?next) &*& nodes(next, tail(vs));

predicate stack(struct Stack *s; list<int> vs) =
    s->head |-> ?h &*& nodes(h, vs);

type I32Predicate = unsafe fn(i32) -> bool;

#[requires(nodes(n, ?vs) &*& forall<int>(vs, p))]
#[ensures(nodes(result, ?filteredVs))]
unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node
    where forall<int>(list<int>, I32Predicate) -> bool
{
    if n == std::ptr::null_mut() {
        std::ptr::null_mut()
    } else {
        open nodes(n, vs);
        bool keep = p((*n).value);
        if (keep) {
            struct Node *next_filtered = filter_nodes((*n).next, p);
            (*n).next = next_filtered;
            close node(n, (*n).value, next_filtered);
            close nodes(n, cons((*n).value, ?tail_filtered));
            return n;
        } else {
            struct Node *next = (*n).next;
            close nodes(next, tail(vs));
            dealloc(n as *mut u8, Layout::new::<Node>());
            return filter_nodes(next, p);
        }
    }
}

#[requires(nodes(n, _) )]
#[ensures(true)]
unsafe fn dispose_nodes(n: *mut Node) {
    if n != std::ptr::null_mut() {
        open nodes(n, ?vs);
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {
    #[requires(true)]
    #[ensures(stack(result, nil))]
    unsafe fn create() -> *mut Stack {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        close stack(stack, nil);
        return stack;
    }

    #[requires(stack(stack, ?vs))]
    #[ensures(stack(stack, cons(value, vs)))]
    unsafe fn push(stack: *mut Stack, value: i32) {
        open stack(stack, vs);
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
        close node(n, value, (*n).next);
        close nodes(n, cons(value, vs));
        close stack(stack, cons(value, vs));
    }

    #[requires(stack(stack, ?vs) &*& vs != nil)]
    #[ensures(stack(stack, tail(vs)) &*& result == head(vs))]
    unsafe fn pop(stack: *mut Stack) -> i32 {
        open stack(stack, vs);
        open nodes((*stack).head, vs);
        let head = (*stack).head;
        int result = (*head).value;
        (*stack).head = (*head).next;
        close stack(stack, tail(vs));
        dealloc(head as *mut u8, Layout::new::<Node>());
        return result;
    }

    #[requires(stack(stack, ?vs) &*& forall<int>(vs, p))]
    #[ensures(stack(stack, ?filteredVs))]
    unsafe fn filter(stack: *mut Stack, p: I32Predicate) {
        open stack(stack, vs);
        struct Node *head = filter_nodes((*stack).head, p);
        (*stack).head = head;
        close stack(stack, filteredVs);
    }

    #[requires(stack(stack, _))]
    #[ensures(true)]
    unsafe fn dispose(stack: *mut Stack) {
        open stack(stack, ?vs);
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

unsafe fn neq_20(x: i32) -> bool {
    x != 20
}

fn main() {
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}